server/gc/src/main/java/org/apache/accumulo/gc/GCRun.java

```diff
@@ -171,8 +179,40 @@ public Stream<Reference> getReferences() {
   }
 
   @Override
-  public Set<TableId> getTableIDs() {
-    return context.getTableIdToNameMap().keySet();
+  public Map<TableId,TableState> getTableIDs() throws InterruptedException {
+    final String tablesPath = context.getZooKeeperRoot() + Constants.ZTABLES;
+    final ZooReader zr = context.getZooReader();
+    int retries = 1;
+    UncheckedIOException ioe = null;
+    while (retries <= 10) {
+      try {
+        zr.sync(tablesPath);
+        final Map<TableId,TableState> tids = new HashMap<>();
+        for (String table : zr.getChildren(tablesPath)) {
+          TableId tableId = TableId.of(table);
+          TableState tableState = null;
+          String statePath = context.getZooKeeperRoot() + Constants.ZTABLES + "/"
+              + tableId.canonical() + Constants.ZTABLE_STATE;
+          byte[] state = zr.getData(statePath, null, null);
```

Comment:
Not completely sure what getData will do if the node does not exist.  If it returns null then ignore the following.  If it throws a NNE then could do the following.

```suggestion
          byte[] state = null;
          try {
             state = zr.getData(statePath, null, null);
           } catch (NoNodeException e) {
               // table was probably deleted after getting list of children... so let it end up as an UNKNOWN state
           }
```

Reply:
I looked at the Accumulo code in ZooReader and the Zookeeper code its calling.  It seems like a KeeperException.NoNode will be thrown when the data node does not exists.

Reply:
That retry at the end seemed more like something for zookeeper errors.  Was thinking the nonode exception is not a zookeeper problem, but a concurrency issue that can  be ignored (no need to redo everything).

Reply:
I think Zookeeper will create the more specific versions of the exception based on the error code.   So you can catch those.  Looked at the ZK source a bit and found the following.

https://github.com/apache/zookeeper/blob/9b6ec9060604347fc996fe69dc33a21222fbd0c4/zookeeper-server/src/main/java/org/apache/zookeeper/ZooKeeper.java#L2028


https://github.com/apache/zookeeper/blob/9b6ec9060604347fc996fe69dc33a21222fbd0c4/zookeeper-server/src/main/java/org/apache/zookeeper/KeeperException.java#L93

Reply:
The ZK javadoc for getData could be improved.

Reply:
Updated in f2cee32fb6117515747872e0b310dde4060702e3 based on your latest information